Simplenote/NoticeController.swift

```diff
@@ -0,0 +1,86 @@
+import UIKit
+
+class NoticeController {
+    static let shared = NoticeController()
+
+    private var notices: [Notice] = []
+    private var current: Notice?
+    private let noticePresenter = NoticePresenter()
+
+    private var activeViewIsBeingTouched: Bool = false
+
+    private init() { }
+
+    func setupNoticeController() {
+        noticePresenter.startListeningToKeyboardNotifications()
+    }
+
+    func present(_ notice: Notice) {
+        if noticePresenter.isPresenting {
+            appendToQueueIfNew(notice)
+            return
+        }
+
+        current = notice
+        let noticeView = makeNoticeView(from: notice)
+
+        noticePresenter.presentNoticeView(noticeView) {
+            if !self.activeViewIsBeingTouched {
+                self.dismiss(noticeView)
+            }
+        }
+    }
+
+    private func dismiss(_ noticeView: NoticeView) {
+        self.noticePresenter.dismissNotification {
+            self.current = nil
+            if !self.notices.isEmpty {
+                self.present(self.notices.removeFirst())
+            }
+        }
+    }
+
+    private func makeNoticeView(from notice: Notice) -> NoticeView {
+        let noticeView: NoticeView = NoticeView.instantiateFromNib()
+        noticeView.message = notice.message
+        noticeView.actionTitle = notice.action?.title
+        noticeView.handler = notice.action?.handler
+        noticeView.delegate = self
+
+        return noticeView
+    }
+
+    /// Confirms if a notice is already contained in notices queue. Appends to queue if new
+    ///
+    private func appendToQueueIfNew(_ notice: Notice) {
+        if notices.contains(notice) {
```

Comment:
We don't really need to guard against this. If the `present` API was invoked multiple times, it would be a bug in the caller side (and shouldn't be handled by the Controller itself).

Same logic applies when using Arrays. When we invoke `array[index]`, if the `index` isn't valid, the system will crash (because the problem lies in the caller's code)

Reply:
I think in this scenario it is easier and less code to handle this here once rather than on the caller's side. 

Let's say we show a notice when you try to log in and forgot to enter a password. You can trigger notice multiple times and there is no way for caller to know that the app is currently showing a notice and which one it is showing.